src/kademlia/dht_tracker.cpp

```diff
@@ -91,95 +91,106 @@ namespace libtorrent { namespace dht {
 		: m_counters(cnt)
 		, m_storage(storage)
 		, m_state(std::move(state))
-		, m_dht(udp::v4(), this, settings, m_state.nid
-			, observer, cnt, m_nodes, storage)
-#if TORRENT_USE_IPV6
-		, m_dht6(udp::v6(), this, settings, m_state.nid6
-			, observer, cnt, m_nodes, storage)
-#endif
 		, m_send_fun(send_fun)
 		, m_log(observer)
 		, m_key_refresh_timer(ios)
-		, m_connection_timer(ios)
-#if TORRENT_USE_IPV6
-		, m_connection_timer6(ios)
-#endif
 		, m_refresh_timer(ios)
 		, m_settings(settings)
-		, m_abort(false)
+		, m_running(false)
 		, m_host_resolver(ios)
 		, m_send_quota(settings.upload_rate_limit)
 		, m_last_tick(aux::time_now())
 	{
 		m_blocker.set_block_timer(m_settings.block_timeout);
 		m_blocker.set_rate_limit(m_settings.block_ratelimit);
+	}
 
-		m_nodes.insert(std::make_pair(m_dht.protocol_family_name(), &m_dht));
-#if TORRENT_USE_IPV6
-		m_nodes.insert(std::make_pair(m_dht6.protocol_family_name(), &m_dht6));
-#endif
-
+	void dht_tracker::update_node_id(aux::session_listen_socket* s)
+	{
+		auto n = m_nodes.find(s);
+		if (n != m_nodes.end())
+			n->second.dht.update_node_id();
 		update_storage_node_ids();
+	}
+
+	void dht_tracker::new_socket(aux::session_listen_socket* s)
+	{
+		address local_address = s->get_local_endpoint().address();
+#if TORRENT_USE_IPV6
+		// don't try to start dht nodes on non-global IPv6 addresses
+		// with IPv4 the interface might be behind NAT so we can't skip them based on the scope of the local address
+		// and we might not have the external address yet
+		if (local_address.is_v6()
+			&& (local_address.to_v6().is_link_local() || local_address.to_v6().is_site_local()))
```

Comment:
Site local addresses are deprecated so it's probably not necessary to check for them.